File: tools/ez_tools/eventlog_parser.py
import os
import pandas as pd
from utils.discovery import find_artifact_files, load_csv_with_progress
from collector.collector import add_rows
from utils.logger import print_and_log 
from utils.filters import print_eventlog_filters
from utils.summary import track_summary

# Channel-specific Event ID filters
EVENT_CHANNEL_FILTERS = {
    "Application": [1000, 1001],
    "Microsoft-Windows-PowerShell/Operational": [4100, 4103, 4104],
    "Microsoft-Windows-RemoteDesktopServices-RdpCoreTS/Operational": [72, 98, 104, 131, 140],
    "Microsoft-Windows-TerminalServices-LocalSessionManager/Operational": [21, 22],
    "Microsoft-Windows-TaskScheduler/Operational": [106, 140, 141, 129, 200, 201],
    "Microsoft-Windows-TerminalServices-RemoteConnectionManager/Operational": [261, 1149],
    "Microsoft-Windows-WinRM/Operational": [169],
    "Security": [1102, 4624, 4625, 4648, 4698, 4702, 4720, 4722, 4723, 4724, 4725, 4726, 4732, 4756],
    "SentinelOne/Operational": [1, 31, 55, 57, 67, 68, 77, 81, 93, 97, 100, 101, 104, 110],
    "System": [7045],
}

def process_eventlog(ez_dir: str, batch_size: int, base_dir: str):
    artifact_name = "EventLogs"
    print_and_log(f"[{artifact_name}] Scanning for relevant CSVs under: {ez_dir}")

    print_eventlog_filters(EVENT_CHANNEL_FILTERS)

    eventlog_files = find_artifact_files(ez_dir, base_dir, artifact_name)

    if not eventlog_files:
        print("[EventLogs] No CSV event log files found.")
        return

    for file_path in eventlog_files:
        print(f"[EventLogs] Processing: {file_path}")
        total_rows = 0
        try:
            for df in load_csv_with_progress(file_path, batch_size, artifact_name=artifact_name):
                filtered_df = df[df.apply(_eventlog_filter, axis=1)]
                timeline_data = []

                for _, row in filtered_df.iterrows():
                    timestamp = row.get("TimeCreated", "")
                    dt = pd.to_datetime(timestamp, utc=True, errors='coerce')
                    if pd.isnull(dt):
                        continue

                    dt_str = dt.isoformat().replace("+00:00", "Z")

                    timeline_row = {
                        "DateTime": dt_str,
                        "TimestampInfo": "Event Time",
                        "ArtifactName": artifact_name,
                        "Tool": "EZ Tools",
                        "Description": row.get("Channel", ""),
                        "DataDetails": row.get("MapDescription", ""),
                        "DataPath": row.get("PayloadData1", ""),
                        "Computer": row.get("Computer", ""),
                        "User": row.get("UserName", ""),
                        "DestinationAddress": row.get("RemoteHost", ""),
                        "EvidencePath": os.path.relpath(file_path, base_dir) if base_dir else file_path,
                        "EventId": row.get("EventId", "")
                    }

                    timeline_data.append(timeline_row)

                total_rows += len(timeline_data)
                add_rows(timeline_data)
                track_summary("EZ Tools", artifact_name, len(timeline_data))
        except Exception as e:
            print(f"[EventLogs] Failed to parse {file_path}: {e}")
            continue

        print_and_log(f"[✓] Parsed {total_rows} timeline rows from: {file_path}")

def _eventlog_filter(row):
    channel = str(row.get("Channel", "")).strip()
    event_id = row.get("EventId")

    if not channel or pd.isnull(event_id):
        return False

    try:
        return int(event_id) in EVENT_CHANNEL_FILTERS.get(channel, [])
    except (ValueError, TypeError):
        return False

File: Tools/EZTools/EventlogParser.cs
using CsvHelper;
using CsvHelper.Configuration;
using ForensicTimeliner.CLI;
using ForensicTimeliner.Interfaces;
using ForensicTimeliner.Models;
using ForensicTimeliner.Utils;
using System.Globalization;

namespace ForensicTimeliner.Tools.EZTools;

public class EventlogParser : IArtifactParser
{
    public List<TimelineRow> Parse(string inputDir, string baseDir, ArtifactDefinition artifact, ParsedArgs args)
    {
        var rows = new List<TimelineRow>();
        Logger.PrintAndLog($"[>] - [{artifact.Artifact}] Scanning for relevant CSVs under: [{inputDir}]", "SCAN");

        var files = Discovery.FindArtifactFiles(inputDir, baseDir, artifact.Artifact);
        if (!files.Any())
        {
            Logger.PrintAndLog($"[!] - [{artifact.Artifact}] No matching files found in: {inputDir}", "WARN");
            return rows;
        }

        foreach (var file in files)
        {
            int parsedRows = 0;
            Logger.PrintAndLog($"[+] - [{artifact.Artifact}] Processing: {Path.GetRelativePath(baseDir, file)}", "PROCESS");

            try
            {
                List<IDictionary<string, object>> filteredRecords = new();

                using (var reader = new StreamReader(file))
                using (var csv = new CsvReader(reader, new CsvConfiguration(CultureInfo.InvariantCulture)
                {
                    HeaderValidated = null,
                    MissingFieldFound = null
                }))
                {
                    foreach (var record in csv.GetRecords<dynamic>())
                    {
                        var dict = (IDictionary<string, object>)record;
                        if (PassesFilter(dict, artifact))
                        {
                            filteredRecords.Add(dict);
                        }
                    }
                }

                ProgressUtils.Show(file, artifact.Artifact, filteredRecords.Count, task =>
                {
                    foreach (var dict in filteredRecords)
                    {
                        var parsedDt = dict.GetDateTime("TimeCreated");
                        if (parsedDt == null) continue;

                        string dtStr = parsedDt.Value.ToString("o").Replace("+00:00", "Z");

                        var row = new TimelineRow
                        {
                            DateTime = dtStr,
                            TimestampInfo = "Event Time",
                            ArtifactName = "Event Logs",
                            Tool = artifact.Tool,
                            Description = dict.GetString("MapDescription"),
                            DataDetails = dict.GetString("Channel"),
                            DataPath = BuildEventlogDataPath(dict),
                            Computer = dict.GetString("Computer"),
                            User = dict.GetString("UserName"),
                            DestinationAddress = dict.GetString("RemoteHost"),
                            EventId = dict.GetString("EventId"),
                            EvidencePath = Path.GetRelativePath(baseDir, file)
                        };

                        if (args.IncludeRawData)
                        {
                            row.RawData = dict.GetString("Payload");
                        }

                        rows.Add(row);
                        parsedRows++;
                        task.Increment(1);
                    }
                });

                Logger.PrintAndLog($"[✓] - [{artifact.Artifact}] Parsed {parsedRows} timeline rows from: {Path.GetFileName(file)}", "SUCCESS");
                LoggerSummary.TrackSummary(artifact.Tool, artifact.Artifact, parsedRows);
            }
            catch (Exception ex)
            {
                Logger.PrintAndLog($"[{artifact.Artifact}] Failed to parse {file}: {ex.Message}", "ERROR");
            }
        }

        return rows;
    }

    private static bool PassesFilter(IDictionary<string, object> dict, ArtifactDefinition artifact)
    {
        if (artifact.IgnoreFilters)
            return true;

        if (!dict.TryGetValue("EventId", out var eventIdObj) || !int.TryParse(eventIdObj?.ToString(), out var eventId))
            return false;

        string channel = dict.GetString("Channel").Trim().ToLowerInvariant();

        // First: Check Event Channel Filters
        var eventFilters = artifact.Filters?.EventChannelFilters;
        if (eventFilters != null)
        {
            var normalized = eventFilters.ToDictionary(
                kvp => kvp.Key.Trim().ToLowerInvariant(),
                kvp => kvp.Value
            );

            if (normalized.TryGetValue(channel, out var validIds) && validIds.Contains(eventId))
                return true;
        }

        // Second: Check Provider Filters
        var providerFilters = artifact.Filters?.ProviderFilters;
        if (providerFilters != null)
        {
            string? provider =
                dict.TryGetValue("Provider", out var p1) ? p1?.ToString() :
                dict.TryGetValue("Event.System.Provider", out var p2) ? p2?.ToString() :
                null;

            if (!string.IsNullOrWhiteSpace(provider))
            {
                var normalized = providerFilters.ToDictionary(
                    kvp => kvp.Key.Trim().ToLowerInvariant(),
                    kvp => kvp.Value
                );

                if (normalized.TryGetValue(provider.Trim().ToLowerInvariant(), out var validIds) && validIds.Contains(eventId))
                    return true;
            }
        }

        // If no filters matched
        return false;
    }

    private static string BuildEventlogDataPath(IDictionary<string, object> dict)
    {
        var parts = new List<string>();

        var channel = dict.GetString("Channel");
        var eventId = dict.GetString("EventID");
        var computer = dict.GetString("Computer");

        if (!string.IsNullOrWhiteSpace(channel)) parts.Add($"Channel: {channel}");
        if (!string.IsNullOrWhiteSpace(eventId)) parts.Add($"EventID: {eventId}");
        if (!string.IsNullOrWhiteSpace(computer)) parts.Add($"Computer: {computer}");

        for (int i = 1; i <= 6; i++)
        {
            var payload = dict.GetString($"PayloadData{i}");
            if (!string.IsNullOrWhiteSpace(payload))
            {
                if (payload.Length > 10000)
                {
                    payload = payload.Substring(0, 10000) + "...[truncated]";
                }
                parts.Add(payload);
            }
        }

        var combined = string.Join(" | ", parts);

        if (!string.IsNullOrEmpty(combined) && !combined.StartsWith(" "))
        {
            combined = " " + combined;
        }

        return combined;
    }
}
